Content.Shared/Damage/DamageSpecifier.cs

```diff
@@ -0,0 +1,257 @@
+using Content.Shared.Damage.Prototypes;
+using Robust.Shared.IoC;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization.Manager.Attributes;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype.Dictionary;
+using Robust.Shared.ViewVariables;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Content.Shared.Damage
+{
+    // TODO DAMAGE UNITS Move this whole class away from, using integers. Also get rid of a lot of the rounding. Just
+    // use DamageUnit math operators.
+
+    /// <summary>
+    ///     Data class with information on a set of damage to deal/heal.
+    /// </summary>
+    /// <remarks>
+    ///     Automatically unpacks damage groups into types, and provides functions to apply resistance sets.
+    ///     Supports basic math operations to modify damage.
+    /// </remarks>
+    [DataDefinition]
+    public class DamageSpecifier
+    {
+        [DataField("types", customTypeSerializer: typeof(PrototypeIdDictionarySerializer<int, DamageTypePrototype>))]
+        private readonly Dictionary<string,int>? _damageTypeDictionary;
+
+        [DataField("groups", customTypeSerializer: typeof(PrototypeIdDictionarySerializer<int, DamageGroupPrototype>))]
+        private readonly Dictionary<string, int>? _damageGroupDictionary;
+
+        /// <summary>
+        ///     Main DamageSpecifier dictionary. Most DamageSpecifier functions exist to somehow modifying this.
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        public Dictionary<string, int> DamageDict
+        {
+            // Cannot use ISerializationHooks.AfterDeserialization(). So instead using this to index damage
+            // prototypes after deserialization only when actually needed.
+            get => GetCombinedDamageDictionary();
+            set => _damageDict = value;
+        }
+        private Dictionary<string, int>? _damageDict;
```

Comment:
Previously I just used `ISerializationHooks.AfterDeserialization()` to initialise  `DamageDict`. The initialisation basically involves taking damage values specified by groups and dividing them into individual damage types. Without this, or some other initialisation function that somehow gets called after all prototypes are loaded in, I'm not sure how best to go about initialising `DamageDict`.

For now, I'm using this caching solution, but hopefully there's a better way to do it. Here, the `GetCombinedDamageDictionary()` function either returns `_damageDict` if it is not null, otherwise it runs through the initialisation steps that were previously in ` ISerializationHooks.AfterDeserialization()`